Added Redis retry_strategy

danihodovic opened the pull request:
case 1. Add a new type definition.
- [ ] checked compilation succeeds with `--target es6` and `--noImplicitAny` options.
- [ ] has correct [naming convention](http://definitelytyped.org/guides/contributing.html#naming-the-file)
- [ ] has a [test file](http://definitelytyped.org/guides/contributing.html#tests) with the suffix of  `-tests.ts` or `-tests.tsx`.

**case 2. Improvement to existing type definition.**
- documentation or source code reference which provides context for the suggested changes.  url http://api.jquery.com/html .
  - it has been reviewed by a DefinitelyTyped member.

Most of the ClientOpts that have to do with reconnecting have been
deprecated in favor of the `retry_strategy` property. The
DefinitelyTyped repo is outdated and is missing the retry_strategy
property. See the docs at https://github.com/NodeRedis/node_redis.

The following options of ClientOpts are deprecated in favor of
retry_strategy:
- retry_max_delay
- connect_timeout
- max_attempts

CodeAnimal commented on redis/redis.d.ts:
```diff
@@ -34,6 +34,10 @@ declare module "redis" {
     }
 
+    interface RetryStrategy {
+      (options: {error: Error, total_retry_time: number, times_connected: number, attempt: number}): number;
```
I'd suggest splitting the options param into a separate interface (e.g. RetryStrategyOptions) for clarity.

Also, the function can potentially return a non-numeric value (e.g. an Error to flush all commands with an individual error, or any other non-numeric value to flush the built in error) ([Documentation here](https://github.com/NodeRedis/node_redis#options-object-properties)). Maybe the return value, at least for now, be `number | Error`? It doesn't cover all scenarios, but it will at least cover most while keeping a reasonable level of type checking.

CodeAnimal commented:
:-1: until [L38 comment](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/10282/files#diff-bcbcaae894e6252f9d088a1529aaa166R38) has been resolved. Almost there though, good work.

danihodovic commented:
> I'd suggest splitting the options param into a separate interface (e.g. RetryStrategyOptions) for clarity.

75320af

> Maybe the return value, at least for now, be number | Error?

75320af

> Also, I can't see any tests added to ensure this change won't be undone or broken in future edits. That would need to be in place, otherwise you couldn't reasonably expect to rely on RetryStrategy to not have breaking changes later.

I'm not sure how throughly testing is done or how tests are organized, but you can have a look at 856e5a2

CodeAnimal commented:
I'd say you're basically there with this :) I'd say the only things missing are the lines I've commented on, but once those are implemented I'd say you've done a good job here :)

Testing is generally done by making sure there are no type-check errors during compile time on the test file(s). So explicitly specifying types are good practise to ensure that the type-check isn't accidentally broken after any changes to the definitions file. This is exactly what you have done in your patch to the tests file, so I think you've got the idea ;)

danihodovic commented:
Addressed the comments in the last two commits.

I'm not sure if using the exported `RedisRetryStrategyOptions` in tests is a good idea since the interface can change in a breaking way and the tests will pass. If the interface is redefined in the tests no undetected breaking change can occur (at the cost of having duplicate logic in tests...).

CodeAnimal commented:
I agree with you about the tests. I would always say steer away from redefining interfaces in tests, so one way around this would be to assign the `options` property values to variables of explicit types within the retry strategy function:

```
function retryStrategyNumber(options: redis.RetryStrategyOptions): number {
  var error: Error = options.error;
  var total_retry_time: number = options.total_retry_time;
  ... etc ...
  return 5000;
}
```

I guess you wouldn't need to do this for both test retry strategy methods, unless one method would dictate differing circumstances that would require future/different type checking, which in this case wouldn't be necessary. It's just a suggestion, when it comes to testing it's hard to say which technique is better/cleaner/more-reusable than another, as long as the full gamut of type definitions are covered, I'd say that the tests have performed their function.

It's up to you to test the options param, in the way you think is best. But I'd say other than that it PR is ready to merge.

:+1:

danihodovic commented:
> I agree with you about the tests. I would always say steer away from redefining interfaces in tests, so one way around this would be to assign the options property values to variables of explicit types within the retry strategy function:

Good idea. Added in f18cd14

> unless one method would dictate differing circumstances that would require further/different type checking

I'm not a Typescript expert, but is it possible to overload functions depending on the return value? Can the two functions provided in the tests have different arguments passed to them? In that case it might make sense to type check the arguments for both functions.
```
function retryStrategyNumber(options: redis.RetryStrategyOptions): number {
  return 5000;
}
function retryStrategyError(options: redis.RetryStrategyOptions): Error {
  return new Error('Foo');
}
```

CodeAnimal commented:
You cannot overload in TypeScript logic like you would in C#. You can create a method with very flexible arguments and then [type guard](https://www.typescriptlang.org/docs/handbook/advanced-types.html#type-guards-and-differentiating-types) those given flexible arguments for further type checking in the code. This is because TypeScript "compiles" down to javascript, in which case the method "signature" would be identical if you had two functions with the same name in the same context (as the argument and return types are not defined in the signature of a javascript method).

But overloading is totally possible in TypeScript's type definition files (as you can see is being utilised even in redis's type defintions file), as no logic is involved, only method signatures for use with TypeScript.

With that in mind, the tests files are only testing for compile time errors (namely, type checking).

> Can the two functions provided in the tests have different arguments passed to them? In that case it might make sense to type check the arguments for both functions.

In theory yes you can pass different arguments, as you could within regular javascript, but as the tests are only really for type checking it is simpler and clearer to implement two methods that return two different value types to use to test. To make the point clear, for example, your methods could have simply returned `null` or no value whatsoever, but the tests would still be the same as long as the signature of the two methods defined the return type. To make the point, the follow method would still be a valid (albeit confusing) test method for a retry strategy method:

```
function retryStrategyNumber(options: redis.RetryStrategyOptions): number {
  return <any>"I'm actually a string, which isn't a number, but I've used type assertion to make the compiler think this could be an acceptable number, making it pass the type checking.";
}
```

So in short, as the definitions file is only declaring method signatures and types, the tests are only checking that given function method signatures and types don't produce compile time errors. So we don't really care about potential "extra" unknown arguments, we only care that the ones defined in our definitions are actually defined with the given types. Logically speaking, if we have two methods that both define a parameter of the type `redis.RetryStrategyOptions`, if we check one parameter's property types, then we have also checked for the other without having to check for them again, as any type checking errors would have been raised during compile time for the first method, checking them again is a little redundant. So, in this case the second method (`retryStrategyError`) is only defined at all solely to check that the return value `Error` doesn't raise a compile time error either.

I hope that all makes sense... that's a lot of words...

danihodovic commented:
That makes sense. Thanks for the write-up.

What is the next step for getting this merged?

CodeAnimal commented:
:+1:  @yuit  This looks good to merge to me.
